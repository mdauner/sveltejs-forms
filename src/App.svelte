<script>
  import {
    Form,
    Input,
    Select,
    Choice,
  } from './components/components.module.js';
  import * as yup from 'yup';

  function handleReset() {
    console.log('form has been reset');
  }

  function handleSubmit({ detail: { values, setSubmitting, resetForm } }) {
    setTimeout(() => {
      console.log(values);
      setSubmitting(false);
      resetForm({
        user: { email: 'test@user.com' },
      });
    }, 2000);
  }

  const schema = yup.object().shape({
    user: yup.object().shape({
      email: yup
        .string()
        .required()
        .email(),
      password: yup.string().min(4),
      language: yup.string().required(),
      os: yup
        .array()
        .of(yup.string().required())
        .min(2),
    }),
  });

  const langOptions = [
    { id: 'svelte', title: 'Svelte' },
    { id: 'react', title: 'React' },
    { id: 'angular', title: 'Angular' },
  ];

  const osOptions = [
    { id: 'macos', title: 'macOS' },
    { id: 'linux', title: 'Linux üêß' },
    { id: 'windows', title: 'Windows' },
  ];
</script>

<style lang="scss">
  :global(.sveltejs-forms) {
    background-color: lightgray;
    display: inline-block;
    padding: 1rem;
    .field {
      margin-bottom: 1rem;
      &.error {
        input,
        select {
          border: 1px solid red;
        }
        .message {
          margin-top: 0.2rem;
          color: red;
          font-size: 0.8rem;
        }
      }
    }
  }
</style>

<Form
  {schema}
  validateOnChange={true}
  validateOnBlur={true}
  on:submit={handleSubmit}
  on:reset={handleReset}
  let:isSubmitting
  let:isValid>
  <Input name="user.email" placeholder="Email" />
  <Input
    name="user.password"
    type="password"
    placeholder="Password"
    multiline />
  <Select name="user.language" options={langOptions} />
  <Choice name="user.os" options={osOptions} multiple />

  <button type="reset">Reset</button>
  <button type="submit" disabled={isSubmitting}>Sign in</button>
  The form is valid: {isValid}
</Form>
