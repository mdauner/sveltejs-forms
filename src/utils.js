export function createObjectWithDefaultValue(defaultValue = '') {
  return new Proxy(
    {},
    {
      get: function(object, property) {
        return Object.prototype.hasOwnProperty.call(object, property)
          ? object[property]
          : defaultValue;
      },
    }
  );
}
