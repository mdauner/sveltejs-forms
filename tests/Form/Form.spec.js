import App from './TestApp.svelte';
import { fireEvent, wait } from '@testing-library/svelte';
import * as yup from 'yup';
import { render } from '../utils';

describe('Form', () => {
  it('click on submit button dispatches submit event and sets isSubmitting to true', async () => {
    const { component, getByText } = await render(App, {
      props: { onSubmit: jest.fn() },
    });
    const signInButton = getByText('Sign in');

    expect(component.form.$$.ctx.$isSubmitting).toBe(false);
    expect(signInButton).not.toHaveAttribute('disabled');

    await fireEvent.click(signInButton);
    await wait();
    expect(component.onSubmit).toHaveBeenCalledTimes(1);
    expect(signInButton).toHaveAttribute('disabled');
    expect(component.form.$$.ctx.$isSubmitting).toBe(true);
  });

  it('onSubmit event returns values, resetForm, setSubmitting', async done => {
    const {
      container,
      component,
      getByText,
      getByPlaceholderText,
    } = await render(App, {
      props: {
        onSubmit: jest.fn(
          ({ detail: { values, setSubmitting, resetForm } }) => {
            expect(values).toMatchSnapshot();

            expect(component.form.$$.ctx.$isSubmitting).toBeTruthy();
            setSubmitting(false);
            expect(component.form.$$.ctx.$isSubmitting).toBeFalsy();

            expect(component.form.$$.ctx.$values).toMatchSnapshot();
            expect(component.form.$$.ctx.$errors).toMatchSnapshot();
            expect(component.form.$$.ctx.$touched).toMatchSnapshot();
            resetForm();
            expect(component.form.$$.ctx.$values).toMatchSnapshot();
            expect(component.form.$$.ctx.$errors).toMatchSnapshot();
            expect(component.form.$$.ctx.$touched).toMatchSnapshot();

            done();
          }
        ),
      },
    });

    const emailInput = getByPlaceholderText('Email');
    await fireEvent.change(emailInput, {
      target: { value: 'test@user.com' },
    });

    const languageSelect = container.querySelector('select');
    await fireEvent.change(languageSelect, {
      target: { value: 'svelte' },
    });

    let osChoice = getByText('macOS');
    await fireEvent.click(osChoice);

    osChoice = getByText('Windows');
    await fireEvent.click(osChoice);

    const signInButton = getByText('Sign in');

    expect(component.form.$$.ctx.$isSubmitting).toBe(false);
    expect(signInButton).not.toHaveAttribute('disabled');

    await fireEvent.click(signInButton);
  });

  it('resetForm resets values to initialValues', async done => {
    const { component, getByText, getByPlaceholderText } = await render(App, {
      props: {
        initialValues: {
          user: { email: 'initial@value.com' },
        },
        onSubmit: jest.fn(({ detail: { resetForm } }) => {
          expect(component.form.$$.ctx.$values.user.email).toEqual(
            'test@user.com'
          );
          resetForm();
          expect(component.form.$$.ctx.$values.user.email).toEqual(
            'initial@value.com'
          );

          done();
        }),
      },
    });

    const emailInput = getByPlaceholderText('Email');
    await fireEvent.change(emailInput, {
      target: { value: 'test@user.com' },
    });

    const signInButton = getByText('Sign in');
    await fireEvent.click(signInButton);
  });

  it('resetForm accepts optional new form data object', async done => {
    const { component, getByText, getByPlaceholderText } = await render(App, {
      props: {
        initialValues: {
          user: { email: 'initial@value.com' },
        },
        onSubmit: jest.fn(({ detail: { resetForm } }) => {
          expect(component.form.$$.ctx.$values.user.email).toEqual(
            'test@user.com'
          );
          resetForm({
            user: { email: 'after@reset.com' },
          });
          expect(component.form.$$.ctx.$values.user.email).toEqual(
            'after@reset.com'
          );

          done();
        }),
      },
    });

    const emailInput = getByPlaceholderText('Email');
    await fireEvent.change(emailInput, {
      target: { value: 'test@user.com' },
    });

    const signInButton = getByText('Sign in');
    await fireEvent.click(signInButton);
  });

  it('shows error message when schema is defined', async () => {
    const schema = yup.object().shape({
      user: yup.object().shape({
        email: yup
          .string()
          .required()
          .email(),
      }),
    });
    const {
      container,
      component,
      getByPlaceholderText,
      getByText,
    } = await render(App, {
      props: { schema },
    });
    const emailInput = getByPlaceholderText('Email');

    await fireEvent.change(emailInput, {
      target: { value: 'invalid value' },
    });
    await fireEvent.blur(emailInput);
    await wait(() => {
      expect(getByText('user.email must be a valid email')).toBeInTheDocument();
    });
    expect(component.form.$$.ctx.$errors).toEqual({
      user: { email: 'user.email must be a valid email' },
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('registers fields and sets default values', async () => {
    const { component } = await render(App);

    expect(component.form.$$.ctx.$values).toMatchObject({
      user: { email: '' },
      language: '',
      os: '',
    });
    expect(component.form.$$.ctx.$touched).toMatchObject({
      user: { email: false },
      language: false,
      os: false,
    });
  });

  it('isValid is undefined initially', async () => {
    const { component } = await render(App);
    expect(component.form.$$.ctx.isValid).toBe(undefined);
  });

  it('sets initial values', async () => {
    const { component } = await render(App, {
      props: { initialValues: { user: { email: 'test@user.com' } } },
    });

    expect(component.form.$$.ctx.$values).toMatchObject({
      user: { email: 'test@user.com' },
      language: '',
      os: '',
    });
  });

  it('matches snapshot', async () => {
    const { container } = await render(App);
    expect(container.firstChild).toMatchSnapshot();
  });
});
