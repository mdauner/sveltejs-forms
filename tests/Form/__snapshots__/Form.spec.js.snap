// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form matches snapshot 1`] = `
<div>
  <form
    class="sveltejs-forms"
  >
    <div
      class="field"
    >
      <input
        name="user.email"
        placeholder="Email"
        type="text"
      />
       
    </div>
     
    <div
      class="field"
    >
      <select
        name="language"
      >
        <option
          value=""
        />
        <option
          value="svelte"
        >
          Svelte
        </option>
        <option
          value="react"
        >
          React
        </option>
        <option
          value="angular"
        >
          Angular
        </option>
      </select>
       
    </div>
     
    <div
      class="field"
    >
      <input
        id="macos"
        name="os"
        type="checkbox"
        value="macos"
      />
       
      <label
        for="macos"
      >
        macOS
      </label>
      
      <input
        id="linux"
        name="os"
        type="checkbox"
        value="linux"
      />
       
      <label
        for="linux"
      >
        Linux üêß
      </label>
      
      <input
        id="windows"
        name="os"
        type="checkbox"
        value="windows"
      />
       
      <label
        for="windows"
      >
        Windows
      </label>
      
       
    </div>
     
    <button
      type="submit"
    >
      Sign in
    </button>
  </form>
</div>
`;

exports[`Form onSubmit event returns values, resetForm, setSubmitting 1`] = `
Object {
  "language": "svelte",
  "os": Array [
    "macos",
    "windows",
  ],
  "user": Object {
    "email": "test@user.com",
  },
}
`;

exports[`Form onSubmit event returns values, resetForm, setSubmitting 2`] = `
Object {
  "language": "svelte",
  "os": Array [
    "macos",
    "windows",
  ],
  "user": Object {
    "email": "test@user.com",
  },
}
`;

exports[`Form onSubmit event returns values, resetForm, setSubmitting 3`] = `Object {}`;

exports[`Form onSubmit event returns values, resetForm, setSubmitting 4`] = `
Object {
  "language": true,
  "os": true,
  "user": Object {
    "email": true,
  },
}
`;

exports[`Form onSubmit event returns values, resetForm, setSubmitting 5`] = `
Object {
  "language": "",
  "os": Array [],
  "user": Object {
    "email": "",
  },
}
`;

exports[`Form onSubmit event returns values, resetForm, setSubmitting 6`] = `Object {}`;

exports[`Form onSubmit event returns values, resetForm, setSubmitting 7`] = `
Object {
  "language": false,
  "os": false,
  "user": Object {
    "email": false,
  },
}
`;

exports[`Form shows error message when schema is defined 1`] = `
<div>
  <form
    class="sveltejs-forms"
  >
    <div
      class="field error"
    >
      <input
        name="user.email"
        placeholder="Email"
        type="text"
      />
       
      <div
        class="message"
      >
        user.email must be a valid email
      </div>
    </div>
     
    <div
      class="field"
    >
      <select
        name="language"
      >
        <option
          value=""
        />
        <option
          value="svelte"
        >
          Svelte
        </option>
        <option
          value="react"
        >
          React
        </option>
        <option
          value="angular"
        >
          Angular
        </option>
      </select>
       
    </div>
     
    <div
      class="field"
    >
      <input
        id="macos"
        name="os"
        type="checkbox"
        value="macos"
      />
       
      <label
        for="macos"
      >
        macOS
      </label>
      
      <input
        id="linux"
        name="os"
        type="checkbox"
        value="linux"
      />
       
      <label
        for="linux"
      >
        Linux üêß
      </label>
      
      <input
        id="windows"
        name="os"
        type="checkbox"
        value="windows"
      />
       
      <label
        for="windows"
      >
        Windows
      </label>
      
       
    </div>
     
    <button
      type="submit"
    >
      Sign in
    </button>
  </form>
</div>
`;
